tests/benchmarks/run-benchmarks.sh

```diff
@@ -117,6 +124,49 @@ hyperfine \
 cat "$RESULT_DIR/startup-time-with-syntax-with-dependencies.md" >> "$REPORT"
 
 
+heading "Startup time with indeterminant syntax"
+hyperfine \
+	"$(printf "%q" "$BAT") --no-config startup-time-src/mystery-file" \
+	--command-name 'bat … mystery-file' \
+	--warmup "$WARMUP_COUNT" \
+    --runs $(($RUN_COUNT * 100)) \
+    --export-markdown "$RESULT_DIR/startup-time-with-indeterminant-syntax.md" \
+    --export-json "$RESULT_DIR/startup-time-with-indeterminant-syntax.json"
+cat "$RESULT_DIR/startup-time-with-indeterminant-syntax.md" >> "$REPORT"
+
+heading "Startup time with manually set syntax"
+hyperfine \
+	"$(printf "%q" "$BAT") --no-config --language=Dockerfile startup-time-src/mystery-file" \
+	--command-name 'bat … --language=Dockerfile mystery-file' \
+	--warmup "$WARMUP_COUNT" \
+    --runs $(($RUN_COUNT * 100)) \
+    --export-markdown "$RESULT_DIR/startup-time-with-manually-set-syntax.md" \
+    --export-json "$RESULT_DIR/startup-time-with-manually-set-syntax.json"
+cat "$RESULT_DIR/startup-time-with-manually-set-syntax.md" >> "$REPORT"
+
+heading "Startup time with mapped syntax"
+hyperfine \
+	"$(printf "%q" "$BAT") --no-config startup-time-src/Containerfile" \
+	--command-name 'bat … Containerfile' \
+	--warmup "$WARMUP_COUNT" \
+    --runs $(($RUN_COUNT * 100)) \
+    --export-markdown "$RESULT_DIR/startup-time-with-mapped-syntax.md" \
+    --export-json "$RESULT_DIR/startup-time-with-mapped-syntax.json"
+cat "$RESULT_DIR/startup-time-with-mapped-syntax.md" >> "$REPORT"
+
+AVG_TIMES=()
+for KIND in indeterminant manually-set mapped; do
+	JSON_SRC="$RESULT_DIR/startup-time-with-$KIND-syntax.json"
+	AVG=$(jq -r '.results[0].mean' "$JSON_SRC")
+	AVG_TIMES+=("$AVG")
+done
+# indeterminant should be slower, because it necessarily has to evaluate all rules
+# to ensure that nothing matches; manually-set and mapped should both be faster
+# because most or all GlobMatcher builds are skipped
+python3 -c "if ${AVG_TIMES[0]} < ${AVG_TIMES[1]}: print('WARN: indeterminant syntax has faster startup than manually set syntax!')"
+python3 -c "if ${AVG_TIMES[0]} < ${AVG_TIMES[2]}: print('WARN: indeterminant syntax has faster startup than mapped syntax!')"
```

Comment:
Is this something we want to keep in here permanently? If so, maybe we could move it to a more specialized benchmarking script? What do you think?

Reply:
Ok, so I did some benchmarks myself. Measuring time differences on the order of ~1 millisecond is challenging. There are two main things that I usually do: you need to work on a quiet machine. Close applications if possible. Definitely close your browser. It makes a huge difference.

Second, you can run a `--shell=none` benchmark if the command is not too fancy (doesn't use any shell syntax, variables, etc). This allows hyperfine to run the command directly, without going through an intermediate shell (and having to correct for that by measuring the shell spawning time).

Finally, for benchmarks you always need to look at the full distribution if the noise level is this high. Sometimes it's more useful to look at the *minimum*, or the *median*, not at the mean — because it's influenced by outliers.

Here are my results using

```
hyperfine \
	--shell=none \
	--warmup 200 \
	--runs 2000 \
	-L version master,2755 \
	--export-json=indeterminant.json \
	'./bat-{version} --no-config mystery-file'

hyperfine \
	--shell=none \
	--warmup 200 \
	--runs 2000 \
	-L version master,2755 \
	--export-json=manually-set.json \
	'./bat-{version} --no-config mystery-file --language=Dockerfile'

hyperfine \
	--shell=none \
	--warmup 200 \
	--runs 2000 \
	-L version master,2755 \
	--export-json=mapped.json \
	'./bat-{version} --no-config Containerfile --language=Dockerfile'
```

and plotted using hyperfine's `plot_histogram.py` script:

indeterminate:

![image](https://github.com/sharkdp/bat/assets/4209276/6acfe0b8-f157-4b1a-b88e-ac6949a09fd6)

manually set:

![image](https://github.com/sharkdp/bat/assets/4209276/24adf3d2-e3bf-4787-a5cf-0ca5b9dbcd51)

mapped:

![image](https://github.com/sharkdp/bat/assets/4209276/cff25460-5551-4d47-a26d-d9a51b9b54d7)


I think it's clear that there is a (statistically) significant advantage!

Reply:
Also: note that bat (and a lot of other programs) change their behavior when the output is piped to /dev/null, which is hyperfine's default setting. For this benchmark, you'll get different results when running with `bat`s `--color=always` or hyperfine's `--output=inherit` option.

They still show an advantage for this branch (see below), except for the "indeterminate" benchmark — where they are on par.

I think those are pretty much the results you expected?

indeterminant:

![image](https://github.com/sharkdp/bat/assets/4209276/e0b52a4f-e1f9-4071-b988-0851524b6ea7)

manually set:

![image](https://github.com/sharkdp/bat/assets/4209276/50683b5c-5e76-4ed5-8bd1-f7bc9591c5df)

![image](https://github.com/sharkdp/bat/assets/4209276/2c60e340-696b-4eef-bcce-f9f821d56e28)

Reply:
This is great to know! Thanks for the help.

> For this benchmark, you'll get different results when running with `bat`s `--color=always` or hyperfine's `--output=inherit` option.

Ah yes good catch. Without `--color=always`, I imagine syntax mappings wouldn't need to be checked at all, which means all 3 benchmarks would behave identically. Coincidentally this is still relevant in verifying that lazy evaluation works, just not what I intended to test.

I've added `--color=always` to these 3 benchmarks, and created another test named (uncreatively) "Startup time without syntax highlighting" as a comparison against "Startup time with syntax highlighting".

> I think those are pretty much the results you expected?

Yes indeed. Just to correct a detail - for the third test (mapped), you shouldn't specify `--language=Dockerfile`. With this option included, you are essentially repeating the second test (manually-set). That being said, I do expect the results to be very similar.

> Is this something we want to keep in here permanently? If so, maybe we could move it to a more specialized benchmarking script? What do you think?

After some thought, I realised this bit of comparison somewhat pointless. After all, the point of this script is to compare across commits, not to compare across different usage scenarios within the same commit. I've decided to remove this section altogether.

Reply:
Out of curiosity, how did you @sharkdp plot this particular style of graph (sorry I don't know its proper name)? I'm trying to benchmark my multithreaded offloading prototype but couldn't find the appropriate option in `plot_histogram.py`.